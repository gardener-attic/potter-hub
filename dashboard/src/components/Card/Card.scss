.Card,
.Card__inner {
  height: 100%;
  width: 100%;
}

.Card {
  margin: 0 1% 1.5em;
}

.Card a,
.Card a:hover {
  // TODO: See why we need important here - https://github.com/bitnami/hex/issues/217
  // sass-lint:disable no-important
  text-decoration: none !important;
}

.Card__Icon img {
  max-height: 64px;
}

/* max 4 items per row (default) */
@media screen and (min-width: 480px) {
  .Card.Card-responsive {
    width: 48%;
  }
}

@media screen and (min-width: 768px) {
  .Card.Card-responsive {
    width: 31%;
  }
}

@media screen and (min-width: 1100px) {
  .Card.Card-responsive {
    width: 23%;
  }
}

/* max 3 items per row */
@media screen and (min-width: 768px) {
  .Card.Card-responsive-3 {
    width: 48%;
  }
}

@media screen and (min-width: 1100px) {
  .Card.Card-responsive-3 {
    width: 31%;
  }
}

/* max 2 items per row */
@media screen and (min-width: 768px) {
  .Card.Card-responsive-2 {
    width: 48%;
  }
}
