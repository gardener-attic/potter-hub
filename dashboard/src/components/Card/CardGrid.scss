.CardGrid {
  display: flex;
  margin: 0 -1%;
  justify-content: space-between;
  flex-wrap: wrap;
}

.CardGrid:after {
  /* Display the last row properly: https://stackoverflow.com/a/34816625 */
  flex: auto;
  content: "";
}
