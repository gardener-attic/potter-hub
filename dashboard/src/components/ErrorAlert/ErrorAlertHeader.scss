.error__icon {
  display: inline-block;
  background: #c7c9c8;
  vertical-align: bottom;
  border-radius: 50%;
  min-height: 30px;
  min-width: 30px;
}

.error__icon > svg {
  margin: auto;
  left: 3px;
  top: 3px;
  color: #ffffff;
  position: relative;
}
