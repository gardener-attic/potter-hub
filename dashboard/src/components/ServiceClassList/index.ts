import ServiceClassList from "./ServiceClassList";

export default ServiceClassList;
