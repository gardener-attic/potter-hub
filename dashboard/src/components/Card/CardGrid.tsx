import * as React from "react";

import "./CardGrid.css";

export interface ICardGridProps {
  className?: string;
  children?: React.ReactChildren | React.ReactNode | string;
}

interface ICardGridState {
  scrollEndIndex: number
}

const SCROLL_THRESHOLD_TOP_PIXEL = 120;
const SCROLL_THRESHOLD_BOTTOM_TO_LOAD = 0.8;
const ITEMS_PER_SCROLL = 32

class CardGrid extends React.Component<ICardGridProps, ICardGridState> {
  constructor(props: any) {
    super(props);
    this.state = {
      scrollEndIndex: ITEMS_PER_SCROLL,
    };
  }

  public componentDidMount() {
    window.addEventListener("scroll", this.onScroll);
  }

  public componentWillUnmount() {
    window.removeEventListener("scroll", this.onScroll)
  }

  public render() {
    return (
      <div className={`CardGrid padding-v-big ${this.props.className || ""}`}>
        {React.Children.toArray(this.props.children).slice(0, this.state.scrollEndIndex)}
      </div>
    );
  }

  private onScroll = (): void => {
    // calculates, if the current scroll position further than the SCROLL_THRESHOLD_BOTTOM_TO_LOAD threshold
    const isScrollAtBottom = window.innerHeight + document.documentElement.scrollTop >
      document.documentElement.offsetHeight * SCROLL_THRESHOLD_BOTTOM_TO_LOAD;

    // calculates, if the current scroll position is above the SCROLL_THRESHOLD_TOP_PIXEL
    const isScrollAtTop = document.documentElement.scrollTop < SCROLL_THRESHOLD_TOP_PIXEL;

    if (isScrollAtBottom && this.state.scrollEndIndex < React.Children.count(this.props.children)) {
      this.setState((state) => ({
        scrollEndIndex: state.scrollEndIndex + ITEMS_PER_SCROLL
      }))
    }

    if (isScrollAtTop && this.state.scrollEndIndex > ITEMS_PER_SCROLL) {
      this.setState((state) => ({
        scrollEndIndex: ITEMS_PER_SCROLL
      }))
    }
  }
}

export default CardGrid;
