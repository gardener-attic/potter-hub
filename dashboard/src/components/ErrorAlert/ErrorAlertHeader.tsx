import * as React from "react";
import { AlertTriangle } from "react-feather";

import "./ErrorAlertHeader.css";

interface IErrorHeaderProps {
  children: string | JSX.Element | Array<string | JSX.Element>;
  icon?: React.Component;
}

class ErrorPageHeader extends React.Component<IErrorHeaderProps> {
  public render() {
    const { children } = this.props;
    const Icon = this.props.icon || AlertTriangle;
    return (
      <header>
        <div className="margin-b-big">
          <h5 className="type-regular">
            <span className="error__icon margin-r-small">
              <Icon />
            </span>
            {children}
          </h5>
        </div>
      </header>
    );
  }
}

export default ErrorPageHeader;
