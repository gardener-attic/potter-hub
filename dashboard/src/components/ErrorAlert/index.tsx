import ErrorAlertHeader from "./ErrorAlertHeader";
import ErrorSelector from "./ErrorSelector";
import MessageAlert from "./MessageAlert";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import ServiceBrokersNotFoundAlert from "./ServiceBrokersNotFoundAlert";
import ServiceCatalogNotInstalledAlert from "./ServiceCatalogNotInstalledAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export {
  ErrorAlertHeader,
  ErrorSelector,
  MessageAlert,
  NotFoundErrorAlert,
  PermissionsErrorAlert,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
  UnexpectedErrorAlert,
};
