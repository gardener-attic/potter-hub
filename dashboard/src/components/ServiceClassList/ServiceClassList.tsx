import * as React from "react";
import { Link } from "react-router-dom";

import { IClusterServiceClass } from "../../shared/ClusterServiceClass";
import { IRBACRole } from "../../shared/types";
import Card, { CardContent, CardFooter, CardGrid, CardIcon } from "../Card";
import { ErrorSelector, MessageAlert } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import PageHeader from "../PageHeader";

export interface IServiceClassListProps {
  error: Error | undefined;
  classes: {
    isFetching: boolean;
    list: IClusterServiceClass[];
  };
  getClasses: () => void;
}

const RequiredRBACRoles: IRBACRole[] = [
  {
    apiGroup: "servicecatalog.k8s.io",
    clusterWide: true,
    resource: "clusterserviceclasses",
    verbs: ["list"],
  },
];

class ServiceClassList extends React.Component<IServiceClassListProps> {
  public componentDidMount() {
    this.props.getClasses();
  }

  public render() {
    const { error, classes } = this.props;
    return (
      <div>
        <PageHeader>
          <h1>Classes</h1>
        </PageHeader>
        <main>
          <LoadingWrapper loaded={!classes.isFetching}>
            {error ? (
              <ErrorSelector
                error={error}
                action="list"
                defaultRequiredRBACRoles={{ list: RequiredRBACRoles }}
                resource="Service Classes"
              />
            ) : (
              <p>Types of services available from all brokers</p>
            )}
            {classes.list.length === 0 ? (
              <MessageAlert level={"warning"}>
                <div>
                  <h5>Service Classes not found.</h5>
                  The Service Catalog server may have failed to populate them. You can find more
                  information about how the Service Catalog works{" "}
                  <a
                    target="_blank"
                    href={
                      "https://kubernetes.io/docs/concepts/extend-kubernetes/service-catalog/#usage"
                    }
                  >
                    here
                  </a>
                  .
                </div>
              </MessageAlert>
            ) : (
              <CardGrid>
                {classes.list
                  .sort((a, b) => a.spec.externalName.localeCompare(b.spec.externalName))
                  .map(svcClass => {
                    const { spec } = svcClass;
                    const { externalMetadata } = spec;
                    const name = externalMetadata
                      ? externalMetadata.displayName
                      : spec.externalName;
                    const description = externalMetadata
                      ? externalMetadata.longDescription
                      : spec.description;
                    const imageUrl = externalMetadata && externalMetadata.imageUrl;
                    const link = `/services/brokers/${svcClass.spec.clusterServiceBrokerName}/classes/${svcClass.spec.externalName}`;

                    const card = (
                      <Card key={svcClass.metadata.uid} responsive={true} responsiveColumns={3}>
                        <CardIcon icon={imageUrl} />
                        <CardContent>
                          <h5>{name}</h5>
                          <p className="margin-b-reset">{description}</p>
                        </CardContent>
                        <CardFooter className="text-c">
                          <Link className="button button-accent" to={link}>
                            Select a plan
                          </Link>
                        </CardFooter>
                      </Card>
                    );
                    return card;
                  })}
              </CardGrid>
            )}
          </LoadingWrapper>
        </main>
      </div>
    );
  }
}

export default ServiceClassList;
