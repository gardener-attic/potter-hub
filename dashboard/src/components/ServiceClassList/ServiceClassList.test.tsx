import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import { IClusterServiceClass } from "../../shared/ClusterServiceClass";
import itBehavesLike from "../../shared/specs";
import { CardGrid } from "../Card";
import { MessageAlert } from "../ErrorAlert";
import ServiceClassList from "./ServiceClassList";

const defaultProps = {
  error: undefined,
  classes: {
    list: [],
    isFetching: false,
  },
  getClasses: jest.fn(),
};

it("should show a warning message if there are no classes available", () => {
  const wrapper = shallow(<ServiceClassList {...defaultProps} />);
  const alert = wrapper.find(MessageAlert);
  expect(alert).toExist();
  expect(alert.html()).toContain("Service Classes not found");
  expect(wrapper).toMatchSnapshot();
});

context("while fetching classes", () => {
  const props = { ...defaultProps, classes: { isFetching: true, list: [] } };

  itBehavesLike("aLoadingComponent", { component: ServiceClassList, props });

  it("matches the snapshot", () => {
    const wrapper = shallow(<ServiceClassList {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a Application header", () => {
    const wrapper = shallow(<ServiceClassList {...props} />);
    expect(wrapper.find("h1").text()).toContain("Classes");
  });
});

context("when there are classes available", () => {
  it("should show a Card item per class", () => {
    const class1 = {
      metadata: {
        uid: "1",
      },
      spec: {
        clusterServiceBrokerName: "azure",
        externalName: "foo",
        description: "this is a service!",
        externalMetadata: {
          imageUrl: "http://foo-image",
        },
      },
    } as IClusterServiceClass;
    const class2 = {
      metadata: {
        uid: "2",
      },
      spec: {
        clusterServiceBrokerName: "gcp",
        externalName: "bar",
        description: "this is a service!",
        externalMetadata: {
          imageUrl: "http://bar-image",
        },
      },
    } as IClusterServiceClass;
    const wrapper = shallow(
      <ServiceClassList
        {...defaultProps}
        classes={{ isFetching: false, list: [class1, class2] }}
      />,
    );
    const cardGrid = wrapper.find(CardGrid);
    expect(cardGrid).toExist();
    expect(cardGrid.children().length).toBe(2);
    expect(wrapper).toMatchSnapshot();
  });
});
