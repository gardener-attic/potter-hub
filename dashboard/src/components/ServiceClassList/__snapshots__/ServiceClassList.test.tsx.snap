// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show a warning message if there are no classes available 1`] = `
<div>
  <PageHeader>
    <h1>
      Classes
    </h1>
  </PageHeader>
  <main>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <p>
        Types of services available from all brokers
      </p>
      <MessageAlertPage
        level="warning"
      >
        <div>
          <h5>
            Service Classes not found.
          </h5>
          The Service Catalog server may have failed to populate them. You can find more information about how the Service Catalog works
           
          <a
            href="https://kubernetes.io/docs/concepts/extend-kubernetes/service-catalog/#usage"
            target="_blank"
          >
            here
          </a>
          .
        </div>
      </MessageAlertPage>
    </LoadingWrapper>
  </main>
</div>
`;

exports[`when there are classes available should show a Card item per class 1`] = `
<div>
  <PageHeader>
    <h1>
      Classes
    </h1>
  </PageHeader>
  <main>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <p>
        Types of services available from all brokers
      </p>
      <CardGrid>
        <Card
          key="2"
          responsive={true}
          responsiveColumns={3}
        >
          <CardIcon
            icon="http://bar-image"
          />
          <CardContent>
            <h5 />
            <p
              className="margin-b-reset"
            />
          </CardContent>
          <CardFooter
            className="text-c"
          >
            <Link
              className="button button-accent"
              to="/services/brokers/gcp/classes/bar"
            >
              Select a plan
            </Link>
          </CardFooter>
        </Card>
        <Card
          key="1"
          responsive={true}
          responsiveColumns={3}
        >
          <CardIcon
            icon="http://foo-image"
          />
          <CardContent>
            <h5 />
            <p
              className="margin-b-reset"
            />
          </CardContent>
          <CardFooter
            className="text-c"
          >
            <Link
              className="button button-accent"
              to="/services/brokers/azure/classes/foo"
            >
              Select a plan
            </Link>
          </CardFooter>
        </Card>
      </CardGrid>
    </LoadingWrapper>
  </main>
</div>
`;

exports[`while fetching classes loading spinner matches the snapshot 1`] = `
<div>
  <PageHeader>
    <h1>
      Classes
    </h1>
  </PageHeader>
  <main>
    <LoadingWrapper
      loaded={false}
      type={0}
    >
      <p>
        Types of services available from all brokers
      </p>
      <MessageAlertPage
        level="warning"
      >
        <div>
          <h5>
            Service Classes not found.
          </h5>
          The Service Catalog server may have failed to populate them. You can find more information about how the Service Catalog works
           
          <a
            href="https://kubernetes.io/docs/concepts/extend-kubernetes/service-catalog/#usage"
            target="_blank"
          >
            here
          </a>
          .
        </div>
      </MessageAlertPage>
    </LoadingWrapper>
  </main>
</div>
`;

exports[`while fetching classes matches the snapshot 1`] = `
<div>
  <PageHeader>
    <h1>
      Classes
    </h1>
  </PageHeader>
  <main>
    <LoadingWrapper
      loaded={false}
      type={0}
    >
      <p>
        Types of services available from all brokers
      </p>
      <MessageAlertPage
        level="warning"
      >
        <div>
          <h5>
            Service Classes not found.
          </h5>
          The Service Catalog server may have failed to populate them. You can find more information about how the Service Catalog works
           
          <a
            href="https://kubernetes.io/docs/concepts/extend-kubernetes/service-catalog/#usage"
            target="_blank"
          >
            here
          </a>
          .
        </div>
      </MessageAlertPage>
    </LoadingWrapper>
  </main>
</div>
`;
